### ISSUE TITLE
Make it easier to recover from "file(s) that should have been pointers, but weren't"

### ISSUE THREAD

**Issue by larsxschneider** [git-lfs/git-lfs#2910]
If files should have been added to Git LFS according to the `.gitattributes` pattern but actually were not added to Git LFS, then users face this kind of message:
```sh
Encountered 1 file(s) that should have been pointers, but weren't:
        test_new.jpg
```

Due to the clean/smudge mechanism, Git will mark these files as modified and it is hard for a Git user to recover from that state.

As [suggested by @ttaylorr](https://github.com/git-lfs/git-lfs/issues/2858#issuecomment-369423130), I would like to teach the migrator to fix this situation with a command like this:
```sh
git lfs migrate import --no-rewrite --include="test_new.jpg"
```

This command would perform the following operation and add a new commit:
```sh
git rm --cached $filepath
git add --force $filepath
git commit -m "Move files properly to GitLFS"
```

This would *not* fix the problem for older commits but it would unblock the user and this is the most important aspect I think. Git LFS would suggest this command if it detects the error like this:

```sh
Encountered 1 file(s) that should have been pointers, but weren't:
        test_new.jpg

Run the following command to fix the problem:
        git lfs migrate import --no-rewrite --include="test_new.jpg"
```

I plan to work on this as soon as I find some time. Please let me know if you see ways to improve this approach. Thanks!

/cc @scval @pauljohn32 @packoman @rgbkrk @alexbbrown @cburanicz @arh-oxbotica

**Comment by pauljohn32** [comment]
I agree with your priority to make it simple and direct to fix user errors that insert binary blobs.

However, can we prevent the problem from happening?  I wish there were a way to stop users who do not have Git LFS installed or enabled from adding and committing binary files. Or pushing would be illegal.  I don't know how to do this, but it seems like you Git gurus ought to be able to do it for us. Maybe this has to be put in Git itself, so it is impossible.

**Comment by rgbkrk** [comment]
What I _truly_ wish for is a way for designers to be able to upload assets directly to GitHub on LFS enabled repos so all they have to do to add a sketch file is drag and drop. I will guarantee-ably have to run this fix each time they contribute.

**Comment by scval** [comment]
Hi @larsxschneider, thank you for clarifying your fix using the migrator and for opening this issue.

I understand that this helps users to circumvent the modified files, but still I do not really like new commits without real modifications.

What about the workaround using `git lfs uninstall && git status && git lfs install`?
This works without a new commit. However, it may lead to confusion if git lfs is installed system-widely and if it is not installed again properly, and it requires adminstrative rights and additional command line switches (--system) if git lfs is installed system-widely. What do you think about a more sophisticated and more convenient version of these commands that does not have these drawbacks? If this would work without adminstrative rights for all cases and with a simple command, this would help us.

**Comment by ttaylorr** [comment]
@pauljohn32 this isn't possible to prevent locally before committing with the current tools provided to us by Git. The best utility we have is a `pre-commit` hook, but since we'd need LFS to install it (i.e., it can't be distributed inside of a repository), having a `pre-commit` implies that LFS is installed, and thus doesn't totally prevent the problem in the first place.

**Comment by ttaylorr** [comment]
> What I _truly_ wish for is a way for designers to be able to upload assets directly to GitHub on LFS enabled repos so all they have to do to add a sketch file is drag and drop.

@rgbkrk forgive me if I'm incorrect, but this seems out of scope for the current discussion. Normally, I'd encourage you to open a new issue on git-lfs/git-lfs, but since this is a GitHub feature request, your best bet would be to email support@github.com.

**Comment by ttaylorr** [comment]
> What about the workaround using `git lfs uninstall && git status && git lfs install`?

@scval I don't know if this would work. I agree that `git lfs uninstall && git status` would mark the working copy as unmodified (in the presenc of a "malformed" LFS pointer), but any checkout of the working copy after running `git lfs install` again would cause the problem to reappear.

**Comment by ttaylorr** [comment]
@larsxschneider I think that this is a _great_ proposal, and I am completely in favor of adding something like this to `git-lfs-migrate(1)`. I envision the following:

1. Upon a fresh checkout (following a `git pull`, `git clone`, `git checkout`, etc), a user encounters a modified working tree because Git LFS expected a pointer, but got a large file from the index.
2. The user runs `git lfs migrate import --include=/path/to/file --no-rewrite`, which adds a _new_ commit to the history (thus, not requiring a force push), updates the currently checked-out ref, and then checks out that new ref.
3. The user observes that their working tree is now unmodified, and they are safe to push and resume work, leaving the large blob still in the history.

If my understanding is correct, I'd like to push forward and get this into v2.5.0, which I'd like to release within the next month.

Here are some questions that I have:

- Should we update the "your working copy is broken" message to include a copy/paste-able command that would perform the above?
- Should we include the new commit in other branches besides the currently checked-out ref? (My thought is that we should not, and instead defer to Git porcelains in to merge this new commit elsewhere).
- For `n > 1` broken files, should we create `n` commits fixing them, or just one? (My preference is one, but I would happily be swayed by a convincing argument).

I have a initial cut of this patch [here](https://gist.github.com/ttaylorr/9a46d1fb5df000e43aae743c018f9a5e), but it obviously needs some work 😄. I'd be happy to work on this myself, or hand it off. Either works.

**Comment by rgbkrk** [comment]
I'd be inclined to one commit fixing them, though I suppose someone could rebase those `n` commits if it goes that way.

**Comment by Hienmaxpro** [comment]
#3052

**Comment by rstecca** [comment]
@larsxschneider 
`git lfs migrate import --no-rewrite --include="path/to/file.jpg"`
results to error
`fatal: expected one or more files with --no-rewrite`

**Comment by ttaylorr** [comment]
Implemented in #3029!

**Comment by nocnokneo** [comment]
What ever happened to the "Run the following command to fix the problem" idea?

**Comment by bk2204** [comment]
It just hasn't been implemented.  We'd be happy to accept a patch.

**Comment by connorjak** [comment]
I'm not sure if this was actually the same issue, but I fixed it by gitignoring all of the problematic files (which didn't hide the change), committing that change, and then un-gitignoring them. I'm not sure if the gitignoring affected it anyway; maybe just committing could work.

**Comment by TaoK** [comment]
I'd like to clarify a couple of the comments just above that made no sense to me, to avoid people needing to navigate multiple other issues:

> @ttaylorr Implemented in #3029!

> @nocnokneo What ever happened to the "Run the following command to fix the problem" idea?

> @bk2204 It just hasn't been implemented. We'd be happy to accept a patch.

I'm not sure what was meant in the last two of these comments, but as I understand it, the now-available command to run to fix this situation, as per the first of these comments, is:

`git lfs migrate import --no-rewrite <problem filename here>`

This will create a new commit in the git repo, with the offending files replaced with pointers as was expected by the complainy process.

**Comment by bk2204** [comment]
The question was about whether we could implement the output that gave the user the exact command to fix the problem.  We have implemented the `git lfs migrate import --no-rewrite` feature, but when you see the "file(s) that should have been pointers, but weren't" message, we don't tell you exactly which invocation to use (yet).

Another easy way to fix this problem is to run `git add --renormalize .` and then commit.  That option is relatively recent in Git, but is a little easier to use if you have it.

**Comment by alist** [comment]
After you get a message like 

```
Encountered 1 file(s) that should have been pointers, but weren't:
        test_new.jpg
```

What's not acceptable about simply `git add .`'ing the changes LFS has made to make these files pointers, then committing?

Is the issue being addressed by `--renormalize` that `add` might otherwise be missing some other should-be-pointer files?

**Comment by alexbbrown** [comment]
Solutions which depend on git add assume the user is in develop / add mode
- is motivated and has permission to change the branches.

The problem I encounter is that checking out and maybe building a series of
branches runs into the error, and fixing it so I can check out the next
branch is time consuming, non obvious and frustrating.

git checkout A
git checkout B
git checkout C
git rebase D

This flow should not encounter blocking errors from git lfs, but does.

**Comment by bk2204** [comment]
Hey, folks,

This is actually a limitation in Git.  Git happens to make this a problem and [it's outlined in the Git FAQ}(https://git-scm.com/docs/gitfaq#Documentation/gitfaq.txt-WhydoIhaveafilethat8217salwaysmodified); it can also happen with other types of smudge and clean filters or working-tree encodings.

There's literally nothing that Git LFS can do to make Git not do this because Git doesn't provide sufficient context for us to know when this case is occurring.  You really do have to fix your repository because otherwise Git will continue to do this.

**Comment by alexbbrown** [comment]
I must still be missing something.  Here's what I imagine happens.

1) I check out
2) at some point before 2: git-lfs noticies that there's a file
which should be filtered, but isn't.  It also presumably knows it has been
checked in (otherwise it wouldn't complain).  It *overwrites the file or
changes the index* thus creating a working or index diff.
3) I try to check out a different branch or refactor.
4) I cannot: some rule in git fires relating to clean branches or
uncommitted changes

in my mind, it's 100% clear that git lfs could execute the following
reasoning.

1) is there a file that's not filtered?
2) is it checked in?
3.1) not it's not checked in.  -> modify it or the index as necessary
3.2) it's checked in already -> don't modify it or the index, but DO warn
the user during git checkin/checkout/rebase ops and in git status

That would 100% avoid the problem I have.

Why would that not work?

-Alex

**Comment by bk2204** [comment]
The way this works is that something tells Git that the file is modified, and Git invokes the clean filter, which is Git LFS.  Git says, "I'd like you to filter this sequence of bytes," and Git LFS says, "Okay," and turns it into a pointer file.

This behavior, which happens during the check out process or other times, is _exactly_ the same process that happens when you run `git add` or `git add --renormalize`, in which case you definitely _do_ want Git LFS to filter it.  You just don't want it automatically filtered at other times, such as when you run `git status`, or `git reset --hard`, but those look exactly like the same situation and are indistinguishable to Git LFS because Git provides no additional information about what it's doing in this situation.

The problem becomes that the index shows the file as being modified if it's a large file checked in as a Git large file instead of an LFS large file, since Git now knows the value in the index that's been hashed is changed.  None of these changes are done by Git LFS, but by Git.  We generally don't touch the index at all, and users would not appreciate us doing so.

The warning that you see from Git LFS happens not at _clean_ time, when the file is being written _into_ the index, but at _smudge_ time, when the files are being checked out from Git and Git is asking Git LFS to turn pointer files into large files.  At that time, we should only ever be seeing pointer files, so we complain if we see a non-pointer file and just return it as is.  That's the point at which we can detect that a user has incorrectly committed a file into the repository, but that's a different invocation from when we get invoked in the way that causes the problem.

**Comment by igorrafael** [comment]
@bk2204 would it be possible to create a command that iterates through every file that it expects to be a pointer and replace with whatever will make the working copy clean?

For example, I would have build machines run this whenever they get the "should have been pointers" error.

**Comment by bk2204** [comment]
That command is `git add --renormalize .` followed by a `git commit`.  Alternatively, you can use `git lfs migrate import --no-rewrite` for this, but the former is a little easier to use, in my opinion.

**Comment by igorrafael** [comment]
Sadly that sometimes is not enough, even paired with `git rm --cached -r .` which I've also been recommended to use.
From my tests, it seems that there is a race condition that can trigger the "should have been pointers" issue exactly by doing the steps you mentioned and then checking out a previous commit. That said, I have not looked any further then getting a consistent repro so I'm not sure what the cause is.
Should I create a new issue or post the repro script here?

**Comment by bk2204** [comment]
If you check out a previous commit that has not been fixed, then you are going to run into this problem and it's not avoidable.  The answer is really to make sure that you don't commit these situations into your codebase in the first place.  You can fix it going forward, but without rewriting history, it's not possible to fix it in old commits.  Again, this is a limitation in how Git works, and it's unfortunate, but not fixable here.

We do eventually intend to add a utility to check this, which we're tracking in #4091, so that people can run a suitable check in their CI environments.

**Comment by igorrafael** [comment]
Thank you, that utility seems to be along the lines of what I was asking for.
I do have one question more question here though: you said you _are_ going to run into this problem while from my tests I see that you _may_ running on this problem and have a script that forces the conditions that can cause that not to happen. Do you really expect it to happen 100% of the time?
I still would like to look into that non-determinism but am now unsure whether it should be reported in this repo (as I need git-lfs for the repro) or on the git mailing list.

**Comment by bk2204** [comment]
If you have a commit where blobs are marked as being `filter=lfs` but they are Git objects, not LFS objects, then this will _likely_ happen, but I cannot guarantee that it will happen in every case.  Whether it happens depends on whether the file is cleaned.  So if you do `find -name '*.git' -prune -o -type f -exec touch '{}' \;`, and then run `git status`, you will likely trigger each broken file, since that forces Git to clean each file since the timestamp has changed.

**Comment by igorrafael** [comment]
That explains the non-determinism I saw, thanks!
